File: project/mobile/lib/models/procedure.dart
/// Procedure model representing a clinical procedure performed on a patient.
class Procedure {
  final int id;
  final int patientId;
  final int? encounterId;
  final int? performedBy;
  final String? performedByName;
  final String? procedureCode;
  final String description;
  final DateTime performedAt;
  final int? durationMinutes;
  final String? outcome;
  final double? baseCost;
  final String? notes;
  final DateTime createdAt;

  Procedure({
    required this.id,
    required this.patientId,
    this.encounterId,
    this.performedBy,
    this.performedByName,
    this.procedureCode,
    required this.description,
    required this.performedAt,
    this.durationMinutes,
    this.outcome,
    this.baseCost,
    this.notes,
    required this.createdAt,
  });

  factory Procedure.fromJson(Map<String, dynamic> json) {
    return Procedure(
      id: json['id'] as int,
      patientId: json['patient_id'] as int,
      encounterId: json['encounter_id'] as int?,
      performedBy: json['performed_by'] as int?,
      performedByName: json['performed_by_name'] as String?,
      procedureCode: json['procedure_code'] as String?,
      description: json['description'] as String,
      performedAt: DateTime.parse(json['performed_at'] as String),
      durationMinutes: json['duration_minutes'] as int?,
      outcome: json['outcome'] as String?,
      baseCost: (json['base_cost'] as num?)?.toDouble(),
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  /// Formatted cost string, e.g. "$1,250.00".
  String get costLabel {
    if (baseCost == null) return 'N/A';
    return '\$${baseCost!.toStringAsFixed(2).replaceAllMapped(
          RegExp(r'(\d)(?=(\d{3})+\.)'),
          (m) => '${m[1]},',
        )}';
  }
}

File: project/backend/app/schemas/procedure.py
"""
Procedure schemas for request/response validation
"""
import bleach
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


def _strip_html(value: Optional[str]) -> Optional[str]:
    if value is None:
        return value
    return bleach.clean(value, tags=[], attributes={}, strip=True)


class ProcedureCreate(BaseModel):
    """Schema for recording a new clinical procedure (Doctor only)."""
    patient_id: int = Field(..., gt=0, description="Patient record ID")
    encounter_id: Optional[int] = Field(None, description="Associated appointment/encounter")
    procedure_code: Optional[str] = Field(None, max_length=20, description="ICD or CPT code")
    description: str = Field(..., min_length=1, description="Description of the procedure")
    performed_at: datetime = Field(..., description="Date and time the procedure was performed")
    duration_minutes: Optional[int] = Field(None, ge=0, description="Procedure duration in minutes")
    outcome: Optional[str] = Field(None, description="Clinical outcome or result")
    base_cost: Optional[float] = Field(None, ge=0, description="List price before insurance")
    notes: Optional[str] = Field(None, description="Free-text clinical notes")

    @field_validator('description', 'outcome', 'notes', mode='before')
    @classmethod
    def sanitize_text(cls, value: Optional[str]) -> Optional[str]:
        return _strip_html(value)


class ProcedureResponse(BaseModel):
    """Full procedure record returned by the API."""
    id: int
    patient_id: int
    encounter_id: Optional[int] = None
    performed_by: Optional[int] = None
    performed_by_name: Optional[str] = None
    procedure_code: Optional[str] = None
    description: str
    performed_at: datetime
    duration_minutes: Optional[int] = None
    outcome: Optional[str] = None
    base_cost: Optional[float] = None
    notes: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ProcedureListResponse(BaseModel):
    """List of procedure records with total count."""
    procedures: List[ProcedureResponse]
    total: int
